File: packages/health_connector_hc_android/lib/src/mappers/measurement_unit_mappers.dart
import 'package:health_connector_core/health_connector_core.dart'
    show
        Mass,
        Energy,
        Length,
        Temperature,
        Pressure,
        Velocity,
        Volume,
        Power,
        BloodGlucose,
        Numeric,
        Percentage,
        RespiratoryRate,
        Vo2Max,
        MeasurementUnit;
import 'package:health_connector_hc_android/src/pigeon/health_connector_platform_api.g.dart'
    show
        MassDto,
        MassUnitDto,
        EnergyDto,
        LengthDto,
        TemperatureDto,
        PressureDto,
        VelocityDto,
        VolumeDto,
        PowerDto,
        BloodGlucoseDto,
        NumericDto,
        PercentageDto,
        MeasurementUnitDto,
        EnergyUnitDto,
        LengthUnitDto,
        TemperatureUnitDto,
        PressureUnitDto,
        VelocityUnitDto,
        VolumeUnitDto,
        PowerUnitDto,
        BloodGlucoseUnitDto,
        NumericUnitDto,
        PercentageUnitDto,
        Vo2MaxDto,
        Vo2MaxUnitDto;
import 'package:meta/meta.dart' show internal;

/// Converts [MeasurementUnit] to [MeasurementUnitDto].
@internal
extension MeasurementUnitToDto on MeasurementUnit {
  MeasurementUnitDto toDto() {
    switch (this) {
      case final Mass unit:
        // Uses kilograms as the transfer unit for consistency.
        return MassDto(value: unit.inKilograms, unit: MassUnitDto.kilograms);
      case final Energy unit:
        // Uses kilocalories as the transfer unit for consistency.
        return EnergyDto(
          value: unit.inKilocalories,
          unit: EnergyUnitDto.kilocalories,
        );
      case final Length unit:
        // Uses meters as the transfer unit for consistency.
        return LengthDto(value: unit.inMeters, unit: LengthUnitDto.meters);
      case final Temperature unit:
        // Uses celsius as the transfer unit for consistency.
        return TemperatureDto(
          value: unit.inCelsius,
          unit: TemperatureUnitDto.celsius,
        );
      case final Pressure unit:
        // Uses millimeters of mercury as the transfer unit.
        return PressureDto(
          value: unit.inMillimetersOfMercury,
          unit: PressureUnitDto.millimetersOfMercury,
        );
      case final Velocity unit:
        // Uses meters per second as the transfer unit for consistency.
        return VelocityDto(
          value: unit.inMetersPerSecond,
          unit: VelocityUnitDto.metersPerSecond,
        );
      case final Volume unit:
        // Uses liters as the transfer unit for consistency.
        return VolumeDto(value: unit.inLiters, unit: VolumeUnitDto.liters);
      case final Power unit:
        // Uses watts as the transfer unit for consistency.
        return PowerDto(value: unit.inWatts, unit: PowerUnitDto.watts);
      case final BloodGlucose unit:
        // Uses millimoles per liter as the transfer unit for consistency.
        return BloodGlucoseDto(
          value: unit.inMillimolesPerLiter,
          unit: BloodGlucoseUnitDto.millimolesPerLiter,
        );
      case final Numeric unit:
        return NumericDto(
          value: unit.value.toDouble(),
          unit: NumericUnitDto.numeric,
        );
      case final Percentage unit:
        // Uses decimal as the transfer unit for consistency.
        return PercentageDto(
          value: unit.asDecimal,
          unit: PercentageUnitDto.decimal,
        );
      case final RespiratoryRate unit:
        return NumericDto(
          value: unit.breathsPerMinute,
          unit: NumericUnitDto.numeric,
        );
      case final Vo2Max unit:
        // Uses milliliters per kilogram per minute as the transfer unit.
        return Vo2MaxDto(
          value: unit.value,
          unit: Vo2MaxUnitDto.millilitersPerKilogramPerMinute,
        );
    }
  }
}

/// Converts [MeasurementUnitDto] to [MeasurementUnit].
@internal
extension MeasurementUnitDtoToDomain on MeasurementUnitDto {
  MeasurementUnit toDomain() {
    switch (this) {
      case final MassDto dto:
        switch (dto.unit) {
          case MassUnitDto.kilograms:
            return Mass.kilograms(dto.value);
          case MassUnitDto.grams:
            return Mass.grams(dto.value);
          case MassUnitDto.pounds:
            return Mass.pounds(dto.value);
          case MassUnitDto.ounces:
            return Mass.ounces(dto.value);
        }
      case final EnergyDto dto:
        switch (dto.unit) {
          case EnergyUnitDto.kilocalories:
            return Energy.kilocalories(dto.value);
          case EnergyUnitDto.kilojoules:
            return Energy.kilojoules(dto.value);
          case EnergyUnitDto.calories:
            return Energy.calories(dto.value);
          case EnergyUnitDto.joules:
            return Energy.joules(dto.value);
        }
      case final LengthDto dto:
        switch (dto.unit) {
          case LengthUnitDto.meters:
            return Length.meters(dto.value);
          case LengthUnitDto.kilometers:
            return Length.kilometers(dto.value);
          case LengthUnitDto.miles:
            return Length.miles(dto.value);
          case LengthUnitDto.feet:
            return Length.feet(dto.value);
          case LengthUnitDto.inches:
            return Length.inches(dto.value);
        }
      case final TemperatureDto dto:
        switch (dto.unit) {
          case TemperatureUnitDto.celsius:
            return Temperature.celsius(dto.value);
          case TemperatureUnitDto.fahrenheit:
            return Temperature.fahrenheit(dto.value);
          case TemperatureUnitDto.kelvin:
            return Temperature.kelvin(dto.value);
        }
      case final PressureDto dto:
        switch (dto.unit) {
          case PressureUnitDto.millimetersOfMercury:
            return Pressure.millimetersOfMercury(dto.value);
        }
      case final VelocityDto dto:
        switch (dto.unit) {
          case VelocityUnitDto.metersPerSecond:
            return Velocity.metersPerSecond(dto.value);
          case VelocityUnitDto.kilometersPerHour:
            return Velocity.kilometersPerHour(dto.value);
          case VelocityUnitDto.milesPerHour:
            return Velocity.milesPerHour(dto.value);
        }
      case final VolumeDto dto:
        switch (dto.unit) {
          case VolumeUnitDto.liters:
            return Volume.liters(dto.value);
          case VolumeUnitDto.milliliters:
            return Volume.milliliters(dto.value);
          case VolumeUnitDto.fluidOuncesUs:
            return Volume.fluidOuncesUs(dto.value);
        }
      case final PowerDto dto:
        switch (dto.unit) {
          case PowerUnitDto.watts:
            return Power.watts(dto.value);
          case PowerUnitDto.kilowatts:
            return Power.kilowatts(dto.value);
        }
      case final BloodGlucoseDto dto:
        switch (dto.unit) {
          case BloodGlucoseUnitDto.millimolesPerLiter:
            return BloodGlucose.millimolesPerLiter(dto.value);
          case BloodGlucoseUnitDto.milligramsPerDeciliter:
            return BloodGlucose.milligramsPerDeciliter(dto.value);
        }
      case final NumericDto dto:
        return Numeric(dto.value);
      case final PercentageDto dto:
        switch (dto.unit) {
          case PercentageUnitDto.decimal:
            return Percentage.fromDecimal(dto.value);
          case PercentageUnitDto.whole:
            return Percentage.fromWhole(dto.value);
        }
      case final Vo2MaxDto dto:
        switch (dto.unit) {
          case Vo2MaxUnitDto.millilitersPerKilogramPerMinute:
            return Vo2Max.millilitersPerKilogramPerMinute(dto.value);
        }
    }
  }
}

File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/mappers/MeasurementUnitMappers.kt
package com.phamtunglam.health_connector_hc_android.mappers

import androidx.health.connect.client.units.BloodGlucose
import androidx.health.connect.client.units.Energy
import androidx.health.connect.client.units.Length
import androidx.health.connect.client.units.Mass
import androidx.health.connect.client.units.Percentage
import androidx.health.connect.client.units.Power
import androidx.health.connect.client.units.Pressure
import androidx.health.connect.client.units.Temperature
import androidx.health.connect.client.units.Velocity
import androidx.health.connect.client.units.Volume
import com.phamtunglam.health_connector_hc_android.pigeon.BloodGlucoseDto
import com.phamtunglam.health_connector_hc_android.pigeon.BloodGlucoseUnitDto
import com.phamtunglam.health_connector_hc_android.pigeon.EnergyDto
import com.phamtunglam.health_connector_hc_android.pigeon.EnergyUnitDto
import com.phamtunglam.health_connector_hc_android.pigeon.LengthDto
import com.phamtunglam.health_connector_hc_android.pigeon.LengthUnitDto
import com.phamtunglam.health_connector_hc_android.pigeon.MassDto
import com.phamtunglam.health_connector_hc_android.pigeon.MassUnitDto
import com.phamtunglam.health_connector_hc_android.pigeon.NumericDto
import com.phamtunglam.health_connector_hc_android.pigeon.NumericUnitDto
import com.phamtunglam.health_connector_hc_android.pigeon.PercentageDto
import com.phamtunglam.health_connector_hc_android.pigeon.PercentageUnitDto
import com.phamtunglam.health_connector_hc_android.pigeon.PowerDto
import com.phamtunglam.health_connector_hc_android.pigeon.PowerUnitDto
import com.phamtunglam.health_connector_hc_android.pigeon.PressureDto
import com.phamtunglam.health_connector_hc_android.pigeon.PressureUnitDto
import com.phamtunglam.health_connector_hc_android.pigeon.TemperatureDto
import com.phamtunglam.health_connector_hc_android.pigeon.TemperatureUnitDto
import com.phamtunglam.health_connector_hc_android.pigeon.VelocityDto
import com.phamtunglam.health_connector_hc_android.pigeon.VelocityUnitDto
import com.phamtunglam.health_connector_hc_android.pigeon.VolumeDto
import com.phamtunglam.health_connector_hc_android.pigeon.VolumeUnitDto

/**
 * Converts a [MassDto] to a Health Connect [Mass] object.
 */
internal fun MassDto.toHealthConnect(): Mass {
    return when (unit) {
        MassUnitDto.KILOGRAMS -> Mass.kilograms(value)
        MassUnitDto.GRAMS -> Mass.grams(value)
        MassUnitDto.POUNDS -> Mass.pounds(value)
        MassUnitDto.OUNCES -> Mass.ounces(value)
    }
}

/**
 * Converts a Health Connect [Mass] object to a [MassDto].
 *
 * Uses kilograms as the transfer unit for consistency.
 */
internal fun Mass.toDto(): MassDto {
    return MassDto(
        value = inKilograms,
        unit = MassUnitDto.KILOGRAMS
    )
}

/**
 * Converts an [EnergyDto] to a Health Connect [Energy] object.
 */
internal fun EnergyDto.toHealthConnect(): Energy {
    return when (unit) {
        EnergyUnitDto.KILOCALORIES -> Energy.kilocalories(value)
        EnergyUnitDto.KILOJOULES -> Energy.kilojoules(value)
        EnergyUnitDto.CALORIES -> Energy.calories(value)
        EnergyUnitDto.JOULES -> Energy.joules(value)
    }
}

/**
 * Converts a Health Connect [Energy] object to an [EnergyDto].
 *
 * Uses kilocalories as the transfer unit for consistency.
 */
internal fun Energy.toDto(): EnergyDto {
    return EnergyDto(
        value = inKilocalories,
        unit = EnergyUnitDto.KILOCALORIES
    )
}

/**
 * Converts a [LengthDto] to a Health Connect [Length] object.
 */
internal fun LengthDto.toHealthConnect(): Length {
    return when (unit) {
        LengthUnitDto.METERS -> Length.meters(value)
        LengthUnitDto.KILOMETERS -> Length.kilometers(value)
        LengthUnitDto.MILES -> Length.miles(value)
        LengthUnitDto.FEET -> Length.feet(value)
        LengthUnitDto.INCHES -> Length.inches(value)
    }
}

/**
 * Converts a Health Connect [Length] object to a [LengthDto].
 *
 * Uses meters as the transfer unit for consistency.
 */
internal fun Length.toDto(): LengthDto {
    return LengthDto(
        value = inMeters,
        unit = LengthUnitDto.METERS
    )
}

/**
 * Converts a [TemperatureDto] to a Health Connect [Temperature] object.
 */
internal fun TemperatureDto.toHealthConnect(): Temperature {
    return when (unit) {
        TemperatureUnitDto.CELSIUS -> Temperature.celsius(value)
        TemperatureUnitDto.FAHRENHEIT -> Temperature.fahrenheit(value)
        TemperatureUnitDto.KELVIN -> Temperature.celsius(value - 273.15)
    }
}

/**
 * Converts a Health Connect [Temperature] object to a [TemperatureDto].
 *
 * Uses celsius as the transfer unit for consistency.
 */
internal fun Temperature.toDto(): TemperatureDto {
    return TemperatureDto(
        value = inCelsius,
        unit = TemperatureUnitDto.CELSIUS
    )
}

/**
 * Converts a [PressureDto] to a Health Connect [Pressure] object.
 */
internal fun PressureDto.toHealthConnect(): Pressure {
    return when (unit) {
        PressureUnitDto.MILLIMETERS_OF_MERCURY -> Pressure.millimetersOfMercury(value)
    }
}

/**
 * Converts a Health Connect [Pressure] object to a [PressureDto].
 *
 * Uses millimeters of mercury as the transfer unit.
 */
internal fun Pressure.toDto(): PressureDto {
    return PressureDto(
        value = inMillimetersOfMercury,
        unit = PressureUnitDto.MILLIMETERS_OF_MERCURY
    )
}

/**
 * Converts a [VelocityDto] to a Health Connect [Velocity] object.
 */
internal fun VelocityDto.toHealthConnect(): Velocity {
    return when (unit) {
        VelocityUnitDto.METERS_PER_SECOND -> Velocity.metersPerSecond(value)
        VelocityUnitDto.KILOMETERS_PER_HOUR -> Velocity.kilometersPerHour(value)
        VelocityUnitDto.MILES_PER_HOUR -> Velocity.milesPerHour(value)
    }
}

/**
 * Converts a Health Connect [Velocity] object to a [VelocityDto].
 *
 * Uses meters per second as the transfer unit for consistency.
 */
internal fun Velocity.toDto(): VelocityDto {
    return VelocityDto(
        value = inMetersPerSecond,
        unit = VelocityUnitDto.METERS_PER_SECOND
    )
}

/**
 * Converts a [VolumeDto] to a Health Connect [Volume] object.
 */
internal fun VolumeDto.toHealthConnect(): Volume {
    return when (unit) {
        VolumeUnitDto.LITERS -> Volume.liters(value)
        VolumeUnitDto.MILLILITERS -> Volume.milliliters(value)
        VolumeUnitDto.FLUID_OUNCES_US -> Volume.fluidOuncesUs(value)
    }
}

/**
 * Converts a Health Connect [Volume] object to a [VolumeDto].
 *
 * Uses liters as the transfer unit for consistency.
 */
internal fun Volume.toDto(): VolumeDto {
    return VolumeDto(
        value = inLiters,
        unit = VolumeUnitDto.LITERS
    )
}

/**
 * Converts a [PowerDto] to a Health Connect [Power] object.
 */
internal fun PowerDto.toHealthConnect(): Power {
    return when (unit) {
        PowerUnitDto.WATTS -> Power.watts(value)
        PowerUnitDto.KILOWATTS -> Power.watts(value * 1000) // Convert kilowatts to watts
    }
}

/**
 * Converts a Health Connect [Power] object to a [PowerDto].
 *
 * Uses watts as the transfer unit for consistency.
 */
internal fun Power.toDto(): PowerDto {
    return PowerDto(
        value = inWatts,
        unit = PowerUnitDto.WATTS
    )
}

/**
 * Converts a [BloodGlucoseDto] to a Health Connect [BloodGlucose] object.
 */
internal fun BloodGlucoseDto.toHealthConnect(): BloodGlucose {
    return when (unit) {
        BloodGlucoseUnitDto.MILLIMOLES_PER_LITER -> BloodGlucose.millimolesPerLiter(value)
        BloodGlucoseUnitDto.MILLIGRAMS_PER_DECILITER -> BloodGlucose.milligramsPerDeciliter(value)
    }
}

/**
 * Converts a Health Connect [BloodGlucose] object to a [BloodGlucoseDto].
 *
 * Uses millimoles per liter as the transfer unit for consistency.
 */
internal fun BloodGlucose.toDto(): BloodGlucoseDto {
    return BloodGlucoseDto(
        value = inMillimolesPerLiter,
        unit = BloodGlucoseUnitDto.MILLIMOLES_PER_LITER
    )
}

/**
 * Converts a [NumericDto] to a numeric value (Long for step counts).
 */
internal fun NumericDto.toLong(): Long {
    return value.toLong()
}

/**
 * Converts a numeric value (Long for step counts) to a [NumericDto].
 *
 * Uses numeric unit for consistency.
 */
internal fun Long.toNumericDto(): NumericDto {
    return NumericDto(
        value = this.toDouble(),
        unit = NumericUnitDto.NUMERIC
    )
}

/**
 * Converts a numeric value (Double for floors climbed) to a [NumericDto].
 *
 * Uses numeric unit for consistency.
 */
internal fun Double.toNumericDto(): NumericDto {
    return NumericDto(
        value = this,
        unit = NumericUnitDto.NUMERIC
    )
}

/**
 * Converts a [PercentageDto] to a Health Connect [Percentage] object.
 */
internal fun PercentageDto.toHealthConnect(): Percentage {
    return when (unit) {
        PercentageUnitDto.DECIMAL -> Percentage(value)
        PercentageUnitDto.WHOLE -> Percentage(value / 100.0)
    }
}

/**
 * Converts a Health Connect [Percentage] object to a [PercentageDto].
 *
 * Uses decimal as the transfer unit for consistency (0.0 to 1.0).
 */
internal fun Percentage.toDto(): PercentageDto {
    return PercentageDto(
        value = value,
        unit = PercentageUnitDto.DECIMAL
    )
}

/**
 * Converts a temperature value (Double in Celsius) to a [TemperatureDto].
 *
 * Uses celsius as the transfer unit for consistency.
 */
internal fun Double.toTemperatureDto(): TemperatureDto {
    return TemperatureDto(
        value = this,
        unit = TemperatureUnitDto.CELSIUS
    )
}

